File: src/ipc/commands.ts
import { invoke } from "@tauri-apps/api/core";
import type { AppSettings, AudioSourcePreset } from "./types";
import { createDefaultSettings, normalizeSettings } from "../settings/settings-store";

export async function pingBackend(): Promise<string> {
  try {
    return await invoke<string>("ping");
  } catch {
    return "frontend-preview";
  }
}

export async function getAppSettings(): Promise<AppSettings> {
  try {
    return normalizeSettings(await invoke<AppSettings>("get_app_settings"));
  } catch {
    return createDefaultSettings();
  }
}

export async function listAudioSourcePresets(): Promise<AudioSourcePreset[]> {
  try {
    return await invoke<AudioSourcePreset[]>("list_audio_source_presets");
  } catch {
    return [];
  }
}

export async function saveAppSettings(settings: AppSettings): Promise<AppSettings> {
  const normalizedSettings = normalizeSettings(settings);

  try {
    return normalizeSettings(await invoke<AppSettings>("save_app_settings", {
      settings: normalizedSettings,
    }));
  } catch {
    return normalizedSettings;
  }
}

export async function setMousePassthrough(enabled: boolean): Promise<boolean> {
  try {
    return await invoke<boolean>("set_mouse_passthrough", { enabled });
  } catch {
    return enabled;
  }
}

export async function openSettingsWindow(): Promise<void> {
  try {
    await invoke("open_settings_window");
  } catch {
    // Preview mode has no Tauri backend.
  }
}

export async function openDebugWindow(): Promise<void> {
  try {
    await invoke("open_debug_window");
  } catch {
    // Preview mode has no Tauri backend.
  }
}

File: src-tauri/src/bridge/commands.rs
use crate::{
    audio::source::{self, AudioSourcePreset},
    bridge::events,
    config::{schema::AppSettings, store},
    platform, window,
};

#[tauri::command]
pub fn ping() -> &'static str {
    "pong"
}

#[tauri::command]
pub fn get_app_settings() -> AppSettings {
    store::load()
}

#[tauri::command]
pub fn list_audio_source_presets() -> Vec<AudioSourcePreset> {
    source::list_source_presets()
}

#[tauri::command]
pub fn save_app_settings(
    app: tauri::AppHandle,
    settings: AppSettings,
) -> Result<AppSettings, String> {
    #[cfg(target_os = "windows")]
    platform::windows::startup::apply(settings.start_with_windows)?;

    store::save(&settings)?;
    window::floating::apply_main_window_layout(&app, &settings)?;
    events::emit_settings_changed(&app, &settings);
    Ok(settings)
}

#[tauri::command]
pub fn set_mouse_passthrough(app: tauri::AppHandle, enabled: bool) -> Result<bool, String> {
    window::floating::set_mouse_passthrough(&app, enabled)?;
    Ok(enabled)
}

#[tauri::command]
pub fn open_settings_window(app: tauri::AppHandle) {
    window::tray::open_settings_window(&app);
}

#[tauri::command]
pub fn open_debug_window(app: tauri::AppHandle) {
    window::tray::open_debug_window(&app);
}
